loky/backend/_thread_pool_limiters.py

```diff
@@ -0,0 +1,481 @@
+        "mkl_win32": {
+            "pattern": "mkl_rt",
+            "api": "mkl"},
+    }
+
+    cls_thread_locals = threading.local()
+
+    def __init__(self):
+        self._load()
+
+    def _load(self):
+        if sys.platform == "darwin":
+            self._modules = self._find_with_clibs_dyld()
+        elif sys.platform == "win32":
+            self._modules = self._find_with_clibs_enum_process_module_ex()
+        else:
+            self._modules = self._find_with_clibs_dl_iterate_phdr()
+
+    def _unload(self):
+        del self._modules
+
+    def _is_supported_clib(self, module_path):
+        module_basename = os.path.basename(module_path).lower()
+        for clib, info in self.SUPPORTED_CLIBS.items():
+            if module_basename.startswith(info['pattern']):
+                return clib, info['api']
+        return None, None
+
+    def _mk_module(self, clib, module_path, api):
+        lib = ctypes.CDLL(module_path)
+        set_func = getattr(lib, self.SUPPORTED_API[api]['set'],
+                           lambda n_thread: NOT_ACCESSIBLE)
+        get_func = getattr(lib, self.SUPPORTED_API[api]['get'],
+                           lambda: NOT_ACCESSIBLE)
+        return dict(name=clib, api=api, module_path=module_path,
+                    lib=lib, version=self.get_version(lib, api),
+                    set=set_func, get=get_func)
+
+    def get_limit(self, name, api, limits):
+        if name in limits:
+            return limits[name]
+        if api in limits:
+            return limits[api]
+        return None
+
+    def set_thread_limits(self, limits=None, apis=None):
+        """Limit maximal number of threads used by supported C-libraries.
+
+        Return a dict of pairs {clib: bool} where bool is True when clib can be
+        dynamically scaled with this function.
+        """
+        if isinstance(limits, int) or limits is None:
+            if apis in ("all", None):
+                apis = self.SUPPORTED_API.keys()
+            elif apis == "blas":
+                apis = ("openblas", "mkl")
+            elif apis == "openmp":
+                apis = ("omp",)
+            else:
+                raise ValueError("apis must be either 'all', 'blas' or "
+                                 "'openmp'. Got {} instead.".format(apis))
+            limits = {api: limits for api in apis}
+        if isinstance(limits, list):
+            limits = {module['name']: module['n_thread'] for module in limits}
+
+        if not isinstance(limits, dict):
+            raise TypeError("limits must either be an int, a dict or None. "
+                            "Got {} instead".format(type(limits)))
+
+        report_threadpool_size = []
+        self._load()
+        for module in self._modules:
+            n_thread = self.get_limit(module['name'], module['api'], limits)
+            set_func = module['set']
+            if n_thread is not None:
+                set_func(openmp_num_threads(n_thread))
```

Comment:
It's weird to use `openmp_num_threads` for non-openmp libraries. I think the wrapper should record the initial value of `num_threads` prior to the first time we set a value and subsequently reuse that original number of thread whenever we set the limit back to None later. This would make it possible to implement a "reset" logic that respects externally defined constraints such as the ones passed by as the user as environment variables such as `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, `OPENBLAS_NUM_THREADS`...

Reply:
Actually I changed my mind: I think we should just not support setting `n_threads=None` in `_set_limits`.

Reply:
Here, `None` is accepted to mean do not do `anything` and we use it when we select the library with `apis`. I think it is fine ot keep it like that but let me know if you think we should disallow passing it, I can check it in the argument before hand.

Reply:
> Here, None is accepted to mean do not do anything and we use it when we select the library with apis.

+1.